trace_decoder/src/typed_mpt.rs

```diff
@@ -0,0 +1,284 @@
+}
+
+impl StateTrie {
+    pub fn insert_by_address(
+        &mut self,
+        address: Address,
+        account: AccountRlp,
+    ) -> Option<Either<H256, AccountRlp>> {
+        self.insert_by_path(TriePath::from_address(address), account)
+    }
+    pub fn insert_by_path(
+        &mut self,
+        path: TriePath,
+        account: AccountRlp,
+    ) -> Option<Either<H256, AccountRlp>> {
+        self.typed.insert(path, account)
+    }
+    pub fn insert_branch_by_path(
+        &mut self,
+        path: TriePath,
+        hash: H256,
+    ) -> Option<Either<H256, AccountRlp>> {
+        self.typed.insert_branch(path, hash)
+    }
+    pub fn get_by_path(&self, path: TriePath) -> Option<Either<H256, AccountRlp>> {
+        self.typed.map.get(&path).copied()
+    }
+    pub fn get_by_address(&self, address: Address) -> Option<Either<H256, AccountRlp>> {
+        self.get_by_path(TriePath::from_hash(keccak_hash::keccak(address)))
+    }
+    pub fn root(&self) -> H256 {
+        self.typed.root()
+    }
+    pub fn iter(&self) -> impl Iterator<Item = (&TriePath, &Either<H256, AccountRlp>)> {
+        self.typed.iter()
+    }
+    pub fn as_hashed_partial_trie(&self) -> mpt_trie::partial_trie::HashedPartialTrie {
+        self.typed.as_hashed_partial_trie()
+    }
+}
+
+impl IntoIterator for StateTrie {
+    type Item = (TriePath, Either<H256, AccountRlp>);
+
+    type IntoIter = Box<dyn Iterator<Item = (TriePath, Either<H256, AccountRlp>)>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.typed.into_iter()
+    }
+}
+
+/// Global, per-account.
+///
+/// See <https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/#storage-trie>
+#[derive(Debug, Clone, Default)]
+pub struct StorageTrie {
+    /// This does NOT use [`TypedMpt`] - T could be anything!
+    map: BTreeMap<TriePath, Either<H256, Vec<u8>>>,
+}
+impl StorageTrie {
+    pub fn insert(&mut self, path: TriePath, value: Vec<u8>) -> Option<Either<H256, Vec<u8>>> {
+        self.map.insert(path, Either::Right(value))
+    }
+    pub fn insert_branch(&mut self, path: TriePath, hash: H256) -> Option<Either<H256, Vec<u8>>> {
+        self.map.insert(path, Either::Left(hash))
+    }
+    pub fn root(&self) -> H256 {
+        self.as_hashed_partial_trie().hash()
+    }
+    pub fn remove(&mut self, path: TriePath) -> Option<Either<H256, Vec<u8>>> {
+        self.map.remove(&path)
+    }
+    pub fn as_hashed_partial_trie(&self) -> mpt_trie::partial_trie::HashedPartialTrie {
+        let mut theirs = mpt_trie::partial_trie::HashedPartialTrie::default();
+        for (k, v) in &self.map {
+            let nibbles = k.into_nibbles();
+            match v {
+                Either::Left(h) => theirs.insert(nibbles, *h),
+                // TODO(0xaatif): why is RLP-encoding here the right thing to do?
+                //                (Our tests fail without it).
+                Either::Right(v) => theirs.insert(nibbles, &*rlp::encode(v)),
```

Comment:
Personal nit: I think it's a bit more readable if we use a specific enum here (eg. `InsertType`) or maybe use two different fields for the types of inserts. I'm good also leaving this though.

Also as for why node values need to be rlp encoded, IIRC it's buried somewhere in the [Ethereum Yellow Paper](https://ethereum.github.io/yellowpaper/paper.pdf) that all trie values are rlp encoded (even for the `txn` & `receipt` tries).

Reply:
I'm using `Either` because we don't have to write `as_left`, `as_right`, `as_ref` etc ourselves - how do you feel about deferring this till the majority of #275 is done, then we can revisit?

Reply:
Yeah that's totally fine! Don't want to block on this.

Reply:
I've rewritten this because I didn't understand tries enough